File: hooks/pr-url-logger.py
#!/usr/bin/env python3
"""
PostToolUse hook: Log PR URL after gh pr create.

When gh pr create succeeds, this hook extracts the PR URL from output
and logs it to LEARNINGS.md for easy reference.
"""

from __future__ import annotations

import json
import re
import sys
from datetime import datetime
from pathlib import Path


def find_learnings_file() -> Path | None:
    """Find .claude/LEARNINGS.md in current project."""
    learnings_path = Path.cwd() / '.claude' / 'LEARNINGS.md'
    return learnings_path if learnings_path.exists() else None


def extract_pr_url(output: str) -> str | None:
    """Extract GitHub PR URL from gh pr create output."""
    # gh pr create outputs URL on success
    # Format: https://github.com/owner/repo/pull/123
    match = re.search(r'https://github\.com/[^/]+/[^/]+/pull/\d+', output)
    return match.group(0) if match else None


def log_pr_to_learnings(pr_url: str, learnings_path: Path) -> bool:
    """Append PR URL to LEARNINGS.md."""
    timestamp = datetime.now().strftime('%Y-%m-%d %H:%M')
    entry = f"\n- [{timestamp}] PR created: {pr_url}\n"

    try:
        with open(learnings_path, 'a', encoding='utf-8') as f:
            f.write(entry)
        return True
    except (IOError, OSError) as e:
        print(f"[PRLogger] Failed to log PR: {e}", file=sys.stderr)
        return False


def main():
    # Read hook input
    try:
        hook_input = json.loads(sys.stdin.read())
    except (json.JSONDecodeError, KeyError):
        sys.exit(0)

    tool_name = hook_input.get('tool_name', '')
    tool_input = hook_input.get('tool_input', {})
    tool_result = hook_input.get('tool_result', {})

    # Only process Bash tool
    if tool_name != 'Bash':
        sys.exit(0)

    # Check if this was a gh pr create command
    command = tool_input.get('command', '')
    if 'gh pr create' not in command:
        sys.exit(0)

    # Get output from tool result
    output = tool_result.get('stdout', '') or tool_result.get('output', '')
    if not output:
        sys.exit(0)

    # Extract PR URL
    pr_url = extract_pr_url(output)
    if not pr_url:
        sys.exit(0)

    # Log to LEARNINGS.md (async - Claude won't wait for this)
    learnings_path = find_learnings_file()
    if learnings_path:
        if log_pr_to_learnings(pr_url, learnings_path):
            print(f"[PRLogger] Logged PR: {pr_url}", file=sys.stderr)

    sys.exit(0)


if __name__ == '__main__':
    main()

File: hooks/pr-url-logger.ts
#!/usr/bin/env bun
/**
 * PostToolUse hook: Log PR URL after gh pr create.
 *
 * When gh pr create succeeds, this hook extracts the PR URL from output
 * and logs it to LEARNINGS.md for easy reference.
 *
 * Port note: this hook NEVER writes to stdout and always exits 0 — there is no deny path. The only
 * observable effects are the appended LEARNINGS.md line and a stderr breadcrumb. A malformed stdin
 * payload must be swallowed (exit 0), exactly as the Python `json.JSONDecodeError` branch did.
 */
import { appendFileSync, existsSync } from "node:fs";
import { join } from "node:path";

/** Find .claude/LEARNINGS.md in current project. */
function findLearningsFile(): string | null {
  const learningsPath = join(process.cwd(), ".claude", "LEARNINGS.md");
  return existsSync(learningsPath) ? learningsPath : null;
}

/** Extract GitHub PR URL from gh pr create output. */
function extractPrUrl(output: string): string | null {
  // gh pr create outputs URL on success
  // Format: https://github.com/owner/repo/pull/123
  const match = output.match(/https:\/\/github\.com\/[^/]+\/[^/]+\/pull\/\d+/);
  return match ? match[0] : null;
}

/** Two-digit zero pad, matching Python's strftime field widths. */
function pad(n: number): string {
  return String(n).padStart(2, "0");
}

/** Append PR URL to LEARNINGS.md. */
function logPrToLearnings(prUrl: string, learningsPath: string): boolean {
  // Local time, minute precision — same as datetime.now().strftime('%Y-%m-%d %H:%M').
  const d = new Date();
  const timestamp =
    `${d.getFullYear()}-${pad(d.getMonth() + 1)}-${pad(d.getDate())} ` +
    `${pad(d.getHours())}:${pad(d.getMinutes())}`;
  const entry = `\n- [${timestamp}] PR created: ${prUrl}\n`;

  try {
    appendFileSync(learningsPath, entry, "utf-8");
    return true;
  } catch (e) {
    console.error(`[PRLogger] Failed to log PR: ${e}`);
    return false;
  }
}

// Read hook input
let hookInput: Record<string, unknown>;
try {
  hookInput = JSON.parse(await Bun.stdin.text());
} catch {
  process.exit(0);
}

const toolName = String(hookInput?.tool_name ?? "");
const toolInput = (hookInput?.tool_input ?? {}) as Record<string, unknown>;
const toolResult = (hookInput?.tool_result ?? {}) as Record<string, unknown>;

// Only process Bash tool
if (toolName !== "Bash") process.exit(0);

// Check if this was a gh pr create command
const command = String(toolInput?.command ?? "");
if (!command.includes("gh pr create")) process.exit(0);

// Get output from tool result (Python `a or b`: empty string falls through to `output`)
const output = String(toolResult?.stdout ?? "") || String(toolResult?.output ?? "");
if (!output) process.exit(0);

// Extract PR URL
const prUrl = extractPrUrl(output);
if (!prUrl) process.exit(0);

// Log to LEARNINGS.md (async - Claude won't wait for this)
const learningsPath = findLearningsFile();
if (learningsPath) {
  if (logPrToLearnings(prUrl, learningsPath)) {
    console.error(`[PRLogger] Logged PR: ${prUrl}`);
  }
}

process.exit(0);
